src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java

```diff
@@ -0,0 +1,432 @@
+/*
+ * Copyright (C) 2020 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.volley.toolbox;
+
+import static com.android.volley.toolbox.NetworkUtility.logSlowRequests;
+
+import android.os.Handler;
+import android.os.Looper;
+import android.os.SystemClock;
+import androidx.annotation.Nullable;
+import com.android.volley.AsyncNetwork;
+import com.android.volley.AuthFailureError;
+import com.android.volley.Cache.Entry;
+import com.android.volley.ClientError;
+import com.android.volley.Header;
+import com.android.volley.NetworkError;
+import com.android.volley.NetworkResponse;
+import com.android.volley.NoConnectionError;
+import com.android.volley.Request;
+import com.android.volley.RetryPolicy;
+import com.android.volley.ServerError;
+import com.android.volley.TimeoutError;
+import com.android.volley.VolleyError;
+import com.android.volley.VolleyLog;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.SocketTimeoutException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+
+/** A network performing Volley requests over an {@link HttpStack}. */
+public class BasicAsyncNetwork extends AsyncNetwork {
+
+    private static final int DEFAULT_POOL_SIZE = 4096;
+
+    /**
+     * @deprecated Should never have been exposed in the API. This field may be removed in a future
+     *     release of Volley.
+     */
+    @Deprecated protected final HttpStack mHttpStack;
+
+    private final BaseHttpStack mBaseHttpStack;
+
+    protected final ByteArrayPool mPool;
+
+    protected ExecutorService mBlockingExecutor;
+
+    protected final Handler mHandler;
+
+    /**
+     * @param httpStack HTTP stack to be used
+     * @deprecated use {@link #BasicAsyncNetwork(BaseHttpStack)} instead to avoid depending on
+     *     Apache HTTP. This method may be removed in a future release of Volley.
+     */
+    @Deprecated
+    public BasicAsyncNetwork(HttpStack httpStack) {
```

Comment:
let's use the Builder pattern here too?

We don't need all these deprecated constructors for a new class. Just don't accept HttpStack from BaseAsyncNetwork.

For that matter, do we actually need to support legacy BaseHttpStacks from BasicAsyncNetwork? Or can we just require that (Basic)Network be used instead in that case?

Reply:
Are you saying our constructor should only take in an AsyncHttpStack? My only worry is then we have to handle both AsyncNetworks and Networks in AsyncRequestQueue. Unless we still require users to pass in an AsyncNetwork into AsyncRequestQueue. Honestly, if you're using a blocking network and blocking stack (and potentially a blocking cache), does it make sense for them to be able to use the AsyncRequestQueue? With the cache is makes sense, since we require a newish api level, but in this case could we just require they use an AsyncNetwork, which requires an AsyncStack?

Reply:
Yes, I agree with what you're saying - I don't see any reason to support the blocking flows for Network in the Async stack. Cache is indeed needed for older devices as you say, but we have an AsyncNetwork implementation that should be usable by anyone willing to use AsyncRequestQueue (or if you don't like Cronet, you could use OkHttp instead, which would also work fine), and if you're using a Network, that broadly defeats the purpose of using the async stack to begin with, and you should just use the legacy one.